File: services/Service.ts

import { GoogleGenAI, Type } from "@google/genai";
import { Gender } from "../types";

/**
 * Stage 1: Condition Map Generation & Styling Analysis
 * Explicitly detecting sleeve length to guide synthesis.
 */
export const analyzeTryOn = async (
  personImageBase64: string, 
  topImageBase64: string | null, 
  bottomImageBase64: string | null,
  dressImageBase64: string | null,
  gender: Gender
) => {
  const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });
  
  const prompt = `
    Act as a Neural Fashion Stylist and Research Engineer.
    TASK: Execute 'Semantic Alignment' and 'Outfit Coordination Analysis'.
    
    1. POSE ESTIMATION: Map the skeletal structure for this ${gender} subject.
    2. SPATIAL BOUNDARIES: Identify where the person is currently wearing clothes.
    3. SLEEVE LENGTH DETECTION: Precisely determine if the PROVIDED reference garments (Top or Dress) have short, half, or long sleeves.
    4. OCCLUSION MAPPING: Identify foreground elements (hands, hair, accessories) that must stay in front of the new garment.
    5. COORDINATION: Suggest matching pants/shoes based on the garment's 'vibe'.
  `;

  const parts = [
    { text: prompt },
    { inlineData: { data: personImageBase64.split(',')[1], mimeType: 'image/jpeg' } }
  ];

  if (topImageBase64) parts.push({ inlineData: { data: topImageBase64.split(',')[1], mimeType: 'image/jpeg' } });
  if (bottomImageBase64) parts.push({ inlineData: { data: bottomImageBase64.split(',')[1], mimeType: 'image/jpeg' } });
  if (dressImageBase64) parts.push({ inlineData: { data: dressImageBase64.split(',')[1], mimeType: 'image/jpeg' } });

  try {
    const response = await ai.models.generateContent({
      model: 'gemini-3-flash-preview',
      contents: { parts },
      config: {
        temperature: 0,
        responseMimeType: "application/json",
        responseSchema: {
          type: Type.OBJECT,
          properties: {
            garmentDescription: { type: Type.STRING },
            personDescription: { type: Type.STRING },
            bodySize: { type: Type.STRING, enum: ['S', 'M', 'L'] },
            technicalPrompt: { type: Type.STRING },
            stylingSuggestions: {
              type: Type.OBJECT,
              properties: {
                suggestedPants: { type: Type.STRING },
                suggestedShoes: { type: Type.STRING },
                suggestedShirt: { type: Type.STRING },
                styleVibe: { type: Type.STRING }
              }
            }
          },
          required: ["garmentDescription", "personDescription", "bodySize", "technicalPrompt", "stylingSuggestions"]
        }
      }
    });

    return JSON.parse(response.text || "{}");
  } catch (error) {
    console.error("Analysis Error:", error);
    throw error;
  }
};

/**
 * Stage 2: Spatial-Aware Detail Preservation Synthesis
 * Updated to respect the sleeve length of the reference garment.
 */
export const generateVirtualTryOnImage = async (
  personImageBase64: string, 
  topImageBase64: string | null, 
  bottomImageBase64: string | null,
  dressImageBase64: string | null,
  technicalDescription: string, 
  bodySize: string, 
  gender: Gender
) => {
  const ai = new GoogleGenAI({ apiKey: process.env.API_KEY });
  
  const parts: any[] = [
    { text: "TEMPLATE: Master image for identity and background." },
    { inlineData: { data: personImageBase64.split(',')[1], mimeType: 'image/jpeg' } }
  ];

  if (topImageBase64) {
    parts.push({ text: "REFERENCE GARMENT: Map this EXACT design onto the person." });
    parts.push({ inlineData: { data: topImageBase64.split(',')[1], mimeType: 'image/jpeg' } });
  }
  if (bottomImageBase64) {
    parts.push({ text: "REFERENCE BOTTOM: Map this EXACT design." });
    parts.push({ inlineData: { data: bottomImageBase64.split(',')[1], mimeType: 'image/jpeg' } });
  }
  if (dressImageBase64) {
    parts.push({ text: "REFERENCE DRESS: Map this EXACT design." });
    parts.push({ inlineData: { data: dressImageBase64.split(',')[1], mimeType: 'image/jpeg' } });
  }

  parts.push({ text: `
    CRITICAL INSTRUCTION: Sleeve Length Fidelity.
    - DO NOT default to the sleeve length of the original clothing in the template.
    - If the REFERENCE garment is short-sleeved (half-hands), you MUST render the person's actual arms/skin below the sleeve.
    - Do NOT stretch the reference garment to cover long sleeves if the reference itself is a t-shirt.
    - Maintain 1:1 texture and logo alignment.
    - Preserve all foreground occlusions (hands, watches, accessories).
    - BLENDING CONTEXT: ${technicalDescription}
  ` });

  try {
    const response = await ai.models.generateContent({
      model: 'gemini-2.5-flash-image',
      contents: { parts }
    });

    for (const part of response.candidates?.[0]?.content?.parts || []) {
      if (part.inlineData) return `data:${part.inlineData.mimeType};base64,${part.inlineData.data}`;
    }
    return null;
  } catch (error) {
    console.error("Synthesis Error:", error);
    throw error;
  }
};

File: services/service.py

import os
import json
import base64
from google import genai
from google.genai import types

def get_client():
    return genai.Client(api_key=os.environ["API_KEY"])

def analyze_try_on(person_b64, top_b64, bottom_b64, dress_b64, gender):
    client = get_client()
    
    prompt = f"""
    Act as a Senior AI Research Engineer specializing in Virtual Try-on synthesis for {gender}.
    TASK: Generate a 'Condition Map' for clothing synthesis.
    Key Analysis Point: 
    - Precisely identify the sleeve length of the provided assets (Short/Half vs Long).
    - Map spatial boundaries and foreground occlusions.
    """
    
    contents = [
        types.Part.from_text(text=prompt),
        types.Part.from_bytes(data=base64.b64decode(person_b64), mime_type="image/jpeg")
    ]
    
    if top_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(top_b64), mime_type="image/jpeg"))
    if bottom_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(bottom_b64), mime_type="image/jpeg"))
    if dress_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(dress_b64), mime_type="image/jpeg"))

    response = client.models.generate_content(
        model='gemini-3-flash-preview',
        contents=contents,
        config=types.GenerateContentConfig(
            temperature=0,
            response_mime_type="application/json",
            response_schema={
                "type": "OBJECT",
                "properties": {
                    "garmentDescription": {"type": "STRING"},
                    "personDescription": {"type": "STRING"},
                    "bodySize": {"type": "STRING"},
                    "technicalPrompt": {"type": "STRING"},
                    "stylingSuggestions": {
                        "type": "OBJECT",
                        "properties": {
                            "suggestedPants": {"type": "STRING"},
                            "suggestedShoes": {"type": "STRING"},
                            "suggestedShirt": {"type": "STRING"},
                            "styleVibe": {"type": "STRING"}
                        }
                    }
                },
                "required": ["garmentDescription", "personDescription", "body_size", "technicalPrompt", "stylingSuggestions"]
            }
        )
    )
    return json.loads(response.text)

def generate_virtual_try_on_image(person_b64, top_b64, bottom_b64, dress_b64, technical_prompt, body_size, gender):
    client = get_client()
    
    prompt = f"""
    TASK: Virtual Try-on Synthesis. 
    Fit provided garments onto the person.
    STRICT RULE: Respect the reference garment's sleeve length. 
    If the reference top is short-sleeved, show the person's arms. 
    Do NOT extend a short-sleeved shirt to match long sleeves from the base template.
    Preserve foreground hands and accessories.
    Context: {technical_prompt}
    """
    
    contents = [
        types.Part.from_text(text=prompt),
        types.Part.from_bytes(data=base64.b64decode(person_b64), mime_type="image/jpeg")
    ]
    
    if top_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(top_b64), mime_type="image/jpeg"))
    if bottom_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(bottom_b64), mime_type="image/jpeg"))
    if dress_b64:
        contents.append(types.Part.from_bytes(data=base64.b64decode(dress_b64), mime_type="image/jpeg"))

    response = client.models.generate_content(
        model='gemini-2.5-flash-image',
        contents=contents,
        config=types.GenerateContentConfig(temperature=0)
    )
    
    for part in response.candidates[0].content.parts:
        if part.inline_data:
            return f"data:{part.inline_data.mime_type};base64,{base64.b64encode(part.inline_data.data).decode()}"
    return None
